csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,350 @@
+      // inputs, we move on to the next ExprGroup. There might be inputs that
+      // have not yet been computed or proven uncomputable; in those cases we
+      // will place them into the vector below so that we can push them onto
+      // the stack so that we can try again.
+      std::vector<ValGroup> uncomputed_producer_groups;
+      for (const ExprGroup& eg : graph_->getDefinitions(vg)) {
+        // NOTE: it suffices to use any Expr* in eg, since they are all
+        // guaranteed to have the same type and identical attributes.
+        Expr* expr = eg->front();
+
+        std::vector<IterDomain*> id_inps;
+        bool all_inputs_computed = true;
+        for (Val* inp : expr->inputs()) {
+          ValGroup vg_inp = graph_->toGroup(inp);
+          auto inp_it = computed_ids.find(vg_inp);
+          if (inp_it != computed_ids.end()) {
+            id_inps.push_back(inp_it->second);
+          } else {
+            // this input is not yet computed
+            all_inputs_computed = false;
+            if (uncomputable_groups.count(vg_inp) == 0) {
+              // Input is not yet proven to be incomputable, so try and
+              // compute it in the next iteration
+              uncomputed_producer_groups.push_back(vg_inp);
+            }
+          }
+        }
+        if (all_inputs_computed) {
+          // Compute new ID expression
+          Expr* id_expr = replayIdExpr(expr, id_inps);
+          // Update the mapping to point to all of the newly created Expr's
+          // outputs
+          NVF_ERROR(id_expr->outputs().size() == expr->outputs().size());
+          for (size_t i : c10::irange(expr->outputs().size())) {
+            ValGroup vg_outp = graph_->toGroup(expr->output((int64_t)i));
+            auto* id_outp = id_expr->output((int64_t)i)->as<IterDomain>();
+            graph_->initializeVal(id_outp, vg_outp);
+            computed_ids.emplace(vg_outp, id_outp);
+          }
+          // No need to look at next ExprGroup
+          break;
+        } else {
+          // Some input is not yet computed
+          if (uncomputed_producer_groups.empty() && !id_inps.empty()) {
+            // The only uncomputed producer groups are uncomputable, but some
+            // are computed. Just pass those computed inputs through. Note that
+            // in this case, id_inps.front() will be the representative of vg
+            // even though it is _not_ actually mapped into that ValGroup.
+            NVF_ERROR(id_inps.size() == 1);
+            computed_ids.emplace(vg, id_inps.front());
+            break;
+          }
+        }
+        if (!uncomputed_producer_groups.empty()) {
+          // Do not look at other ExprGroups before we try computing these
+          // uncomputed producer groups
+          break;
+        }
+      }
+
+      if (uncomputed_producer_groups.empty()) {
+        // All of the defining expressions are proven uncomputable, so mark vg
```

Comment:
Why is this the case? We are still within the while loop of `eval_stack`. Suppose `vg` has all inputs computed, so nothing would be added to `uncomputed_producer_groups`. Am I missing something?

Reply:
I rearranged the code slightly to make it more clear what condition I'm asserting on.